Change representation of PeerId to CIDv1

JerryHue opened the pull request:
Addresses #2259.

To encapsulate both errors from Multihash and CID, we change the interface of `from_bytes` to return a `PeerIdError` in the case of `Err`.

JerryHue committed:
Change representation of PeerId to CIDv1

To encapsulate both errors from Multihash and CID,
we change the interface of `from_bytes` to
return a `PeerIdError` in the case of `Err`.

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
+    ///
+    /// If the CID's multihash does not fit the specifications:
+    ///
+    /// * use a valid hashing algorithm for peer IDs, or
```
```suggestion
    /// * use an  invalid hashing algorithm for peer IDs, or
```

Correct?

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
+    ///
+    /// * is a CIDv0 with a multihash that fits the specifications
+    /// for multihash, or
+    /// * is a CIDv1 with the proper codec (LIBP2P_KEY_CODEC),
```
I am confused by this. Shouldn't the opposite be written, along the lines of: "If XXX does not fit the specification, ..."?

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
+    /// Returns a raw bytes representation of the `PeerId`.
+    pub fn to_cidv1_bytes(&self) -> Vec<u8> {
+        self.cid.to_bytes()
+    }
```
Is this needed anywhere? If not, I would prefer removing it to ease future breaking changes. If I am not mistaken test logic should be able to call `self.cid.to_bytes` directly.

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
         bs58::encode(self.to_bytes()).into_string()
     }
 
+    /// Returns a `base32` encoded string of this `PeerId`.
```
```suggestion
    /// Returns a `base32` multibase encoded string of this `PeerId`.
```

mxinden replied:
```suggestion
    /// Returns a `base32` encoded string of this `PeerId`.
    ///
    /// This corresponds to the new peer ID [representation].
    ///
    /// [representation]: https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md
```

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
+        self.cid.to_bytes()
+    }
+
+    /// Returns a base-58 raw encoded string of this `PeerId`.
```
```suggestion
    /// Returns a base-58 raw (non-multibase) encoded string of this `PeerId`.
```

mxinden replied:
```suggestion
    /// Returns a base-58 raw encoded string of this `PeerId`.
    ///
    /// This corresponds to the old peer ID [representation].
    ///
    /// [representation]: https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md
```

mxinden commented on core/src/peer_id.rs:
```diff
@@ -77,44 +102,86 @@ impl PeerId {
+
+    /// Returns a base-58 raw encoded string of this `PeerId`.
     pub fn to_base58(&self) -> String {
         bs58::encode(self.to_bytes()).into_string()
```
The following would fail given that `<PeerId as FromStr>` expects the string to start with `Qm` or `1`, right?

```rust
assert!(cidv1_peer_id == PeerId::from_str(cidv1_peer_id.to_base58()));
```

I am not quite sure how to proceed in this case. Maybe convert the CIDv1 peer ID to the old format first?

mxinden replied:
> Maybe convert the CIDv1 peer ID to the old format first?

What do you think @JerryHue?

mxinden commented on core/src/peer_id.rs:
```diff
@@ -164,21 +231,68 @@ impl From<PeerId> for Vec<u8> {
+    CidError(CidError),
+}
+
+impl fmt::Display for PeerIdError {
```
`libp2p-core` already imports `thiserror`. What do you think of using it instead of this manual implementation?

mxinden reviewed (commented):
First off, very solid work! Thanks @JerryHue.

I have a couple of comments below:

mxinden commented on core/src/peer_id.rs:
```diff
@@ -28,12 +30,15 @@ use thiserror::Error;
 ///
-/// The data is a multihash of the public key of the peer.
+/// The data is a cid of the public key of the peer.
 #[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
```
If I am not mistaken the implementation of `Eq` is simply derived on both `PeerId` and `Cid`. Given that we store all `PeerID`s as CIDv1s now, an old style `PeerId` and a new style `PeerId` both parsed from a `String` would still equal, correct? Could you add a test to make sure this is the case going forward?

JerryHue committed:
Fix omission of `multibase` on docs

Co-authored-by: Max Inden <mail@max-inden.de>

JerryHue committed:
Add detailed description for `to_base58`

Co-authored-by: Max Inden <mail@max-inden.de>

JerryHue committed:
Improve doc for `from_cid`

JerryHue committed:
Add test to check old PeerId equals new PeerId

We check whether the old representation of
a PeerId and the new representation of a
PeerId produce the same PeerId.
